File: typescript/tests/unit/logger.test.ts
import { expect } from "chai";
import * as sinon from "sinon";
import {
  log,
  logError,
  logDebug,
  logWarn,
  logInfo,
  logAPICall,
  logAPIResponseWithDetails,
  maskSensitiveData,
  setLogLevel,
  getLogLevel,
} from "../../src/utils/logger";

describe("Logger", () => {
  let stdoutWriteStub: sinon.SinonStub;
  let stderrWriteStub: sinon.SinonStub;

  beforeEach(() => {
    stdoutWriteStub = sinon.stub(process.stdout, "write");
    stderrWriteStub = sinon.stub(process.stderr, "write");
    setLogLevel('DEBUG');
  });

  afterEach(() => {
    sinon.restore();
    setLogLevel('INFO');
  });

  describe("Basic Logging Functions", () => {
    it("should log message to stdout", () => {
      log("test message");
      expect(stdoutWriteStub.called).to.be.true;
      expect(stdoutWriteStub.firstCall.args[0]).to.include("test message");
    });

    it("should log info message with INFO prefix", () => {
      logInfo("info message");
      expect(stdoutWriteStub.called).to.be.true;
      const output = stdoutWriteStub.firstCall.args[0];
      expect(output).to.include("ℹ️  INFO");
      expect(output).to.include("info message");
    });

    it("should log debug message with DEBUG prefix", () => {
      logDebug("debug message");
      expect(stdoutWriteStub.called).to.be.true;
      const output = stdoutWriteStub.firstCall.args[0];
      expect(output).to.include("🐛 DEBUG");
      expect(output).to.include("debug message");
    });

    it("should log warning message with WARN prefix to stderr", () => {
      logWarn("warning message");
      expect(stderrWriteStub.called).to.be.true;
      const output = stderrWriteStub.firstCall.args[0];
      expect(output).to.include("⚠️  WARN");
      expect(output).to.include("warning message");
    });

    it("should log error message to stderr with ERROR prefix", () => {
      logError("error message");
      expect(stderrWriteStub.called).to.be.true;
      const output = stderrWriteStub.firstCall.args[0];
      expect(output).to.include("❌ ERROR");
      expect(output).to.include("error message");
    });

    it("should log error with Error object stack trace", () => {
      const error = new Error("test error");
      logError("operation failed", error);
      expect(stderrWriteStub.called).to.be.true;
      const fullOutput = stderrWriteStub.getCalls().map(c => c.args[0]).join("");
      expect(fullOutput).to.include("test error");
      expect(fullOutput).to.include("Stack");
    });
  });

  describe("API Logging Functions", () => {
    it("should log API call with API name", () => {
      logAPICall("CreateSession");
      expect(stdoutWriteStub.called).to.be.true;
      const output = stdoutWriteStub.firstCall.args[0];
      expect(output).to.include("🔗 API Call");
      expect(output).to.include("CreateSession");
    });

    it("should log API response with success status", () => {
      logAPIResponseWithDetails("CreateSession", "req-123", true, {
        session_id: "sess-456",
        resource_url: "https://example.com",
      });
      expect(stdoutWriteStub.called).to.be.true;
      const output = stdoutWriteStub.getCalls().map(c => c.args[0]).join("");
      expect(output).to.include("✅ API Response");
      expect(output).to.include("CreateSession");
      expect(output).to.include("req-123");
      expect(output).to.include("session_id=sess-456");
      expect(output).to.include("resource_url=https://example.com");
    });

    it("should log API response with error status", () => {
      logAPIResponseWithDetails("DeleteSession", "req-789", false, {}, "Not found");
      expect(stderrWriteStub.called).to.be.true;
      const output = stderrWriteStub.getCalls().map(c => c.args[0]).join("");
      expect(output).to.include("❌ API Response Failed");
      expect(output).to.include("DeleteSession");
    });
  });

  describe("Sensitive Data Masking", () => {
    it("should mask API keys in objects", () => {
      const data = {
        api_key: "sk-1234567890abcdef",
        username: "testuser",
      };
      const masked = maskSensitiveData(data);
      expect(masked.api_key).to.equal("sk****ef");
      expect(masked.username).to.equal("testuser");
    });

    it("should mask tokens in objects", () => {
      const data = {
        access_token: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9",
        user_id: "12345",
      };
      const masked = maskSensitiveData(data);
      expect(masked.access_token).to.equal("ey****J9");
      expect(masked.user_id).to.equal("12345");
    });

    it("should mask passwords in objects", () => {
      const data = {
        password: "mySecurePassword123",
        email: "user@example.com",
      };
      const masked = maskSensitiveData(data);
      expect(masked.password).to.equal("my****23");
      expect(masked.email).to.equal("user@example.com");
    });

    it("should recursively mask nested objects", () => {
      const data = {
        user: {
          name: "John",
          api_key: "secret123",
        },
        credentials: {
          password: "pass456",
        },
      };
      const masked = maskSensitiveData(data);
      expect(masked.user.api_key).to.equal("se****23");
      expect(masked.credentials.password).to.equal("pa****56");
      expect(masked.user.name).to.equal("John");
    });

    it("should mask arrays of objects", () => {
      const data = [
        { api_key: "key123", name: "item1" },
        { api_key: "key456", name: "item2" },
      ];
      const masked = maskSensitiveData(data);
      expect(masked[0].api_key).to.equal("ke****23");
      expect(masked[1].api_key).to.equal("ke****56");
    });

    it("should handle custom sensitive fields", () => {
      const data = {
        custom_secret: "mysecretvalue",
        public_field: "publicvalue",
      };
      const masked = maskSensitiveData(data, ["custom_secret"]);
      expect(masked.custom_secret).to.equal("my****ue");
      expect(masked.public_field).to.equal("publicvalue");
    });
  });


  describe("Multiple Arguments Logging", () => {
    it("should log message with multiple arguments", () => {
      log("test message", { key: "value" }, "additional");
      expect(stdoutWriteStub.called).to.be.true;
      expect(stdoutWriteStub.callCount).to.be.greaterThan(1);
    });

    it("should format objects with proper indentation", () => {
      log("message", { nested: { field: "value" } });
      const calls = stdoutWriteStub.getCalls();
      expect(calls.some(c => c.args[0].includes(JSON.stringify({ nested: { field: "value" } }, null, 2)))).to.be.true;
    });
  });

  describe("Edge Cases", () => {
    it("should handle null values gracefully", () => {
      expect(() => log("message", null)).to.not.throw();
    });

    it("should handle undefined values gracefully", () => {
      expect(() => log("message", undefined)).to.not.throw();
    });

    it("should handle circular references in masking", () => {
      const data: any = { api_key: "secret", ref: null };
      data.ref = data;
      expect(() => maskSensitiveData(data)).to.not.throw();
    });

    it("should handle very long strings", () => {
      const longString = "a".repeat(10000);
      expect(() => log(longString)).to.not.throw();
    });
  });

  describe("Log Level Control", () => {
    it("should filter DEBUG logs when log level is INFO", () => {
      setLogLevel('INFO');
      logDebug("debug message");
      expect(stdoutWriteStub.called).to.be.false;
    });

    it("should show DEBUG logs when log level is DEBUG", () => {
      setLogLevel('DEBUG');
      logDebug("debug message");
      expect(stdoutWriteStub.called).to.be.true;
    });

    it("should filter lower level logs when log level is ERROR", () => {
      setLogLevel('ERROR');
      logInfo("info message");
      expect(stdoutWriteStub.called).to.be.false;
      logDebug("debug message");
      expect(stdoutWriteStub.called).to.be.false;
    });

    it("should show error logs regardless of level", () => {
      setLogLevel('ERROR');
      logError("error message");
      expect(stderrWriteStub.called).to.be.true;
    });

    it("should get current log level", () => {
      setLogLevel('DEBUG');
      expect(getLogLevel()).to.equal('DEBUG');
      setLogLevel('INFO');
      expect(getLogLevel()).to.equal('INFO');
    });

    it("should filter INFO logs when log level is WARN", () => {
      setLogLevel('WARN');
      logInfo("info message");
      expect(stdoutWriteStub.called).to.be.false;
    });

    it("should show WARN logs when log level is WARN", () => {
      setLogLevel('WARN');
      logWarn("warning message");
      expect(stderrWriteStub.called).to.be.true;
    });
  });

  describe("Output Stream Correctness", () => {
    it("INFO should write to stdout", () => {
      logInfo("info");
      expect(stdoutWriteStub.called).to.be.true;
      expect(stderrWriteStub.called).to.be.false;
    });

    it("WARN should write to stderr", () => {
      logWarn("warning");
      expect(stderrWriteStub.called).to.be.true;
      expect(stdoutWriteStub.called).to.be.false;
    });

    it("ERROR should write to stderr", () => {
      logError("error");
      expect(stderrWriteStub.called).to.be.true;
      expect(stdoutWriteStub.called).to.be.false;
    });

    it("DEBUG should write to stdout", () => {
      logDebug("debug");
      expect(stdoutWriteStub.called).to.be.true;
      expect(stderrWriteStub.called).to.be.false;
    });

  });

  describe("API Logging with Log Level", () => {
    it("should respect log level for API logging", () => {
      setLogLevel('ERROR');
      logAPICall("TestAPI");
      expect(stdoutWriteStub.called).to.be.false;
    });

    it("should log API calls at INFO level", () => {
      setLogLevel('INFO');
      logAPICall("TestAPI");
      expect(stdoutWriteStub.called).to.be.true;
    });
  });
});

File: golang/pkg/agentbay/logger_test.go
package agentbay

import (
	"encoding/json"
	"strings"
	"testing"
)

// TestLogAPICall verifies LogAPICall produces correct formatted output
func TestLogAPICall(t *testing.T) {
	tests := []struct {
		name    string
		apiName string
		params  string
	}{
		{
			name:    "basic API call",
			apiName: "CreateSession",
			params:  "ImageId=linux_latest, IsVpc=true",
		},
		{
			name:    "empty params",
			apiName: "ListSessions",
			params:  "",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Just verify it doesn't panic
			LogAPICall(tt.apiName, tt.params)
		})
	}
}

// TestLogAPIResponseWithDetailsSuccess verifies successful API response logging
func TestLogAPIResponseWithDetailsSuccess(t *testing.T) {
	keyFields := map[string]interface{}{
		"session_id":   "sess_123456",
		"resource_url": "https://example.com/session/sess_123456",
		"is_vpc":       true,
	}

	responseBody := map[string]interface{}{
		"success": true,
		"data": map[string]interface{}{
			"session_id": "sess_123456",
		},
	}
	fullResponse, _ := json.Marshal(responseBody)

	// Verify it doesn't panic with valid inputs
	LogAPIResponseWithDetails("CreateSession", "req_12345", true, keyFields, string(fullResponse))
}

// TestLogAPIResponseWithDetailsFailure verifies failed API response logging
func TestLogAPIResponseWithDetailsFailure(t *testing.T) {
	responseBody := map[string]interface{}{
		"success": false,
		"code":    "SessionNotFound",
		"message": "Session not found",
	}
	fullResponse, _ := json.Marshal(responseBody)

	// Verify it doesn't panic with error response
	LogAPIResponseWithDetails("GetSession", "req_67890", false, nil, string(fullResponse))
}

// TestLogOperationErrorWithoutStack verifies error logging without stack trace
func TestLogOperationErrorWithoutStack(t *testing.T) {
	// Verify it doesn't panic
	LogOperationError("CreateSession", "Network timeout", false)
}

// TestLogOperationErrorWithStack verifies error logging with stack trace
func TestLogOperationErrorWithStack(t *testing.T) {
	// Verify it doesn't panic
	LogOperationError("DeleteSession", "Permission denied", true)
}

// TestMaskSensitiveDataWithMapApiKey verifies API key masking in maps
func TestMaskSensitiveDataWithMapApiKey(t *testing.T) {
	data := map[string]interface{}{
		"api_key": "sk-1234567890abcdef",
		"user":    "john@example.com",
	}

	masked := MaskSensitiveData(data).(map[string]interface{})

	// Verify API key is masked (first 2 + **** + last 2)
	if masked["api_key"] != "sk****ef" {
		t.Errorf("expected 'sk****ef', got '%v'", masked["api_key"])
	}

	// Verify non-sensitive data is unchanged
	if masked["user"] != "john@example.com" {
		t.Errorf("expected 'john@example.com', got '%v'", masked["user"])
	}
}

// TestMaskSensitiveDataWithPassword verifies password masking
func TestMaskSensitiveDataWithPassword(t *testing.T) {
	data := map[string]interface{}{
		"password": "mysecretpassword123",
		"username": "admin",
	}

	masked := MaskSensitiveData(data).(map[string]interface{})

	// Verify password is masked (first 2 + **** + last 2)
	if masked["password"] != "my****23" {
		t.Errorf("expected 'my****23', got '%v'", masked["password"])
	}

	// Verify username is unchanged
	if masked["username"] != "admin" {
		t.Errorf("expected 'admin', got '%v'", masked["username"])
	}
}

// TestMaskSensitiveDataWithToken verifies token masking
func TestMaskSensitiveDataWithToken(t *testing.T) {
	data := map[string]interface{}{
		"access_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9",
		"user_id":      "user_12345",
	}

	masked := MaskSensitiveData(data).(map[string]interface{})

	// Verify token is masked
	if !strings.Contains(masked["access_token"].(string), "****") {
		t.Errorf("token should be masked, got '%v'", masked["access_token"])
	}

	// Verify user_id is unchanged
	if masked["user_id"] != "user_12345" {
		t.Errorf("expected 'user_12345', got '%v'", masked["user_id"])
	}
}

// TestMaskSensitiveDataWithNestedStructure verifies recursive masking
func TestMaskSensitiveDataWithNestedStructure(t *testing.T) {
	data := map[string]interface{}{
		"user": map[string]interface{}{
			"name":     "John",
			"password": "secret123",
		},
		"api_key": "key_abcdef123456",
	}

	masked := MaskSensitiveData(data).(map[string]interface{})
	userMasked := masked["user"].(map[string]interface{})

	// Verify nested password is masked (first 2 + **** + last 2)
	if userMasked["password"] != "se****23" {
		t.Errorf("expected 'se****23', got '%v'", userMasked["password"])
	}

	// Verify nested name is unchanged
	if userMasked["name"] != "John" {
		t.Errorf("expected 'John', got '%v'", userMasked["name"])
	}

	// Verify top-level api_key is masked (first 2 + **** + last 2)
	if masked["api_key"] != "ke****56" {
		t.Errorf("expected 'ke****56', got '%v'", masked["api_key"])
	}
}

// TestMaskSensitiveDataWithArray verifies masking in arrays
func TestMaskSensitiveDataWithArray(t *testing.T) {
	data := []interface{}{
		map[string]interface{}{
			"api_key": "key_111111",
			"name":    "item1",
		},
		map[string]interface{}{
			"api_key": "key_222222",
			"name":    "item2",
		},
	}

	masked := MaskSensitiveData(data).([]interface{})

	// Verify first item's api_key is masked (first 2 + **** + last 2)
	first := masked[0].(map[string]interface{})
	if first["api_key"] != "ke****11" {
		t.Errorf("expected 'ke****11', got '%v'", first["api_key"])
	}

	// Verify second item's api_key is masked (first 2 + **** + last 2)
	second := masked[1].(map[string]interface{})
	if second["api_key"] != "ke****22" {
		t.Errorf("expected 'ke****22', got '%v'", second["api_key"])
	}
}

// TestMaskSensitiveDataWithShortSecret verifies masking of short sensitive values
func TestMaskSensitiveDataWithShortSecret(t *testing.T) {
	data := map[string]interface{}{
		"secret": "abc",   // Less than 5 chars
		"token":  "12345", // Exactly 5 chars
	}

	masked := MaskSensitiveData(data).(map[string]interface{})

	// Verify short secret is masked completely
	if masked["secret"] != "****" {
		t.Errorf("expected '****', got '%v'", masked["secret"])
	}

	// Verify 5-char token is masked (first 2 + **** + last 2)
	if masked["token"] != "12****45" {
		t.Errorf("expected '12****45', got '%v'", masked["token"])
	}
}

// TestMaskSensitiveDataPreservesNonSensitive verifies non-sensitive data is unchanged
func TestMaskSensitiveDataPreservesNonSensitive(t *testing.T) {
	data := map[string]interface{}{
		"user_id":      12345,
		"session_id":   "sess_abc123",
		"resource_url": "https://example.com",
		"total_count":  100,
		"is_vpc":       true,
	}

	masked := MaskSensitiveData(data).(map[string]interface{})

	// Verify all non-sensitive data is unchanged
	tests := []struct {
		key      string
		expected interface{}
	}{
		{"user_id", 12345},
		{"session_id", "sess_abc123"},
		{"resource_url", "https://example.com"},
		{"total_count", 100},
		{"is_vpc", true},
	}

	for _, tt := range tests {
		if masked[tt.key] != tt.expected {
			t.Errorf("expected %v for key %s, got %v", tt.expected, tt.key, masked[tt.key])
		}
	}
}

// TestIsSensitiveField verifies the sensitive field detection
func TestIsSensitiveField(t *testing.T) {
	tests := []struct {
		field    string
		expected bool
	}{
		{"api_key", true},
		{"apikey", true},
		{"api-key", true},
		{"password", true},
		{"passwd", true},
		{"token", true},
		{"access_token", true},
		{"secret", true},
		{"session_id", false},
		{"user_name", false},
		{"resource_url", false},
		{"Authorization", true},
		{"AUTHORIZATION", true},
	}

	for _, tt := range tests {
		result := isSensitiveField(tt.field, SENSITIVE_FIELDS)
		if result != tt.expected {
			t.Errorf("isSensitiveField(%q) = %v, want %v", tt.field, result, tt.expected)
		}
	}
}

// TestLogLevelControl tests the log level control functionality
func TestLogLevelControl(t *testing.T) {
	// Save original log level
	originalLevel := GetLogLevel()
	defer SetLogLevel(originalLevel)

	// Test 1: Default log level is INFO
	if GetLogLevel() != LOG_INFO {
		t.Errorf("Expected default log level to be LOG_INFO, got %d", GetLogLevel())
	}

	// Test 2: Set log level to DEBUG
	SetLogLevel(LOG_DEBUG)
	if GetLogLevel() != LOG_DEBUG {
		t.Errorf("Expected log level to be LOG_DEBUG, got %d", GetLogLevel())
	}

	// Test 3: Set log level to WARN
	SetLogLevel(LOG_WARN)
	if GetLogLevel() != LOG_WARN {
		t.Errorf("Expected log level to be LOG_WARN, got %d", GetLogLevel())
	}

	// Test 4: Set log level to ERROR
	SetLogLevel(LOG_ERROR)
	if GetLogLevel() != LOG_ERROR {
		t.Errorf("Expected log level to be LOG_ERROR, got %d", GetLogLevel())
	}

	// Test 5: Invalid log level should not change
	SetLogLevel(999)
	if GetLogLevel() != LOG_ERROR {
		t.Errorf("Expected log level to remain LOG_ERROR after invalid set, got %d", GetLogLevel())
	}

	// Test 6: Reset to DEBUG and verify LogAPICall output
	t.Log("\n--- Test 6: LogAPICall at DEBUG level (should print Request) ---")
	SetLogLevel(LOG_DEBUG)
	LogAPICall("TestAPI", `{"param":"value"}`)

	// Test 7: Set to INFO and verify LogAPICall output
	t.Log("\n--- Test 7: LogAPICall at INFO level (Request should not print) ---")
	SetLogLevel(LOG_INFO)
	LogAPICall("TestAPI", `{"param":"value"}`)

	// Test 8: LogAPIResponseWithDetails at DEBUG level
	t.Log("\n--- Test 8: LogAPIResponseWithDetails at DEBUG level (Full Response should print) ---")
	SetLogLevel(LOG_DEBUG)
	keyFields := map[string]interface{}{
		"session_id": "session-12345",
		"status":     "running",
	}
	LogAPIResponseWithDetails("CreateSession", "req-123", true, keyFields, `{"status":200,"data":{"id":"123"}}`)

	// Test 9: LogAPIResponseWithDetails at INFO level
	t.Log("\n--- Test 9: LogAPIResponseWithDetails at INFO level (Full Response should not print) ---")
	SetLogLevel(LOG_INFO)
	LogAPIResponseWithDetails("CreateSession", "req-123", true, keyFields, `{"status":200,"data":{"id":"123"}}`)

	// Test 10: LogAPIResponseWithDetails at WARN level (should not print anything)
	t.Log("\n--- Test 10: LogAPIResponseWithDetails at WARN level (nothing should print) ---")
	SetLogLevel(LOG_WARN)
	LogAPIResponseWithDetails("CreateSession", "req-123", true, keyFields, `{"status":200,"data":{"id":"123"}}`)

	t.Log("\n--- All log level control tests completed ---")
}

// TestLogLevelConstants verifies log level constants
func TestLogLevelConstants(t *testing.T) {
	if LOG_DEBUG != 0 {
		t.Errorf("Expected LOG_DEBUG to be 0, got %d", LOG_DEBUG)
	}
	if LOG_INFO != 1 {
		t.Errorf("Expected LOG_INFO to be 1, got %d", LOG_INFO)
	}
	if LOG_WARN != 2 {
		t.Errorf("Expected LOG_WARN to be 2, got %d", LOG_WARN)
	}
	if LOG_ERROR != 3 {
		t.Errorf("Expected LOG_ERROR to be 3, got %d", LOG_ERROR)
	}
}
